Source/ARTMessage.h

```diff
@@ -16,4 +17,11 @@ NS_ASSUME_NONNULL_BEGIN
 
 @end
 
+@interface ARTMessage (Decryption)
+
++ (instancetype)fromEncodedJsonObject:(NSDictionary *)json channelOptions:(ARTChannelOptions *)options error:(NSError **)error;
```

Comment:
Not sure if dictionary is a right format here, `ably-java` allows the JSON to be passed as `String` and `JsonObject` types. I think if they're saved or retrieved outside of Ably SDK, they're probably stored as Strings, so it may be a good idea to think about implementing an alternative definition of this method?

Reply:
> it may be a good idea to think about implementing an alternative definition of this method?

or, change the `NSDictionary` to universal type `id` and check type inside the function body.

Reply:
What about its `NSArray` variant? Should it be an array of strings or dictionaries or else? Or to leave only one function that figures out whether input is an array or not and returns an array or single object. That would also be confusing. I've just used the same types that an actual parser already uses - `ARTJsonLikeEncoder`. `NSDictionary` translates into `[String: Any]` in Swift and it's pretty conventional type for a JSON. WDYT @QuintinWillison @ikurek @lukasz-szyszkowski @lawrence-forooghian

Reply:
Ok, ably-java just checks if an element of an array is a JSON object - https://github.com/ably/ably-java/blob/382829ee6df70875d372162bb84186344e2f7dfb/lib/src/main/java/io/ably/lib/types/Message.java#L243

Reply:
It depends how universal this method should be. Do we want to have one method for both `String` and `Dictionary`? Or maybe it's a better idea to have two separate methods for these types. IMO universal type aka `id` is a bad idea because there is no simple information what we should pass as an input.

Reply:
From TP4 (emphasis mine):

> `fromEncoded` and `fromEncodedArray` are alternative constructors that take an _(already deserialized)_ `Message`-like object (or array of such objects)

and from the [IDL](https://docs.ably.com/client-lib-development-guide/features/#idl):

```
+fromEncoded(JsonObject, ChannelOptions?) -> Message
```

> `JSONObject` and `JSONArray` denote any type or interface in the target language that represents an [RFC4627](https://www.ietf.org/rfc/rfc4627.txt) object or array value respectively.

The spec doesn't explicitly define the meaning of "deserialized", but from what I can tell, it means "already decoded by a JSON or MessagePack parser".

So, I think we don't need to worry about accepting a string here, seems like we only need to accept `NSDictionary` and `NSArray`.

Reply:
That's going to create some complexity for Dart SDK. Right now I simply pass the String to `ably-java` and leave the decoding to native implementation. If we decide to use `NSDictionary` for `ably-cocoa` I'll need to deserialize the object on Dart side and pass either `NSDictionary` or `JSONObject` to platform implementations. It's possible, but it will be some additional work on Dart side, and can be a little bit tricky because `JSONObject` comes from GSON lib and I'm not sure if I'll be Able to use it in Dart bridge 🤔

Reply:
As of these methods already exist in ably-java I've created them here as well for consistency. Probably we need to add them to specs too. WDYT @QuintinWillison  @lawrence-forooghian @lukasz-szyszkowski 
https://github.com/ably/ably-cocoa/pull/1307/commits/ee20ff4781733fe25882e74c854a6fa477c801b8

Reply:
In Java the `fromEncoded` method takes `String messageJson` as parameter, for MessagePack there's an alternative method `fromMsgpack`. There's also a `fromEncoded` variant that takes `JsonObject` but I don't think it's a good idea, because `JsonObject` is provided by `GSON` library, so this requires users to rely on external sources for serialization

Reply:
@ikurek So what was the progress on that? Can I now remove methods accepting string as an argument?